benchmark/rocblas/blas3/gemm.cpp

```diff
@@ -91,81 +103,120 @@ void run(benchmark::State& state, rocblas_handle& rb_handle, int t_a_i,
+      a_size, reinterpret_cast<rocm_scalar_t*>(a.data()));
+  blas_benchmark::utils::HIPVector<rocm_scalar_t> b_gpu(
+      b_size, reinterpret_cast<rocm_scalar_t*>(b.data()));
+  blas_benchmark::utils::HIPVector<rocm_scalar_t> c_gpu(
+      c_size, reinterpret_cast<rocm_scalar_t*>(c.data()));
 
-#ifdef BLAS_VERIFY_BENCHMARK
-    // Reference gemm
-    std::vector<scalar_t> c_ref = c;
-    reference_blas::gemm(t_a_str, t_b_str, m, n, k, alpha, a.data(), lda,
-                         b.data(), ldb, beta, c_ref.data(), ldc);
+  constexpr const bool is_half = std::is_same_v<scalar_t, cl::sycl::half>;
 
-    // Rocblas verification gemm
-    std::vector<scalar_t> c_temp = c;
-    {
-      blas_benchmark::utils::HIPVector<scalar_t, true> c_temp_gpu(
-          c_size, c_temp.data());
-      rocblas_gemm_f<scalar_t>(rb_handle, trans_a_rb, trans_b_rb, m, n, k,
-                               &alpha, a_gpu, lda, b_gpu, ldb, &beta,
-                               c_temp_gpu, ldc);
-    }
+  rocm_scalar_t alpha_rocm, beta_rocm;
 
-    std::ostringstream err_stream;
-    if (!utils::compare_vectors(c_temp, c_ref, err_stream, "")) {
-      const std::string& err_str = err_stream.str();
-      state.SkipWithError(err_str.c_str());
-      *success = false;
-    };
+  if constexpr (is_half) {
+#ifdef BLAS_ENABLE_HALF
+    // sycl::half to rocblas__half
+    alpha_rocm = *reinterpret_cast<rocm_scalar_t*>(&alpha);
+    beta_rocm = *reinterpret_cast<rocm_scalar_t*>(&beta);
+  } else {
 #endif
-    auto blas_warmup = [&]() -> void {
-      rocblas_gemm_f<scalar_t>(rb_handle, trans_a_rb, trans_b_rb, m, n, k,
-                               &alpha, a_gpu, lda, b_gpu, ldb, &beta, c_gpu,
-                               ldc);
-      return;
-    };
-
-    hipEvent_t start, stop;
-    CHECK_HIP_ERROR(hipEventCreate(&start));
-    CHECK_HIP_ERROR(hipEventCreate(&stop));
-
-    auto blas_method_def = [&]() -> std::vector<hipEvent_t> {
-      CHECK_HIP_ERROR(hipEventRecord(start, NULL));
-      rocblas_gemm_f<scalar_t>(rb_handle, trans_a_rb, trans_b_rb, m, n, k,
-                               &alpha, a_gpu, lda, b_gpu, ldb, &beta, c_gpu,
-                               ldc);
-      CHECK_HIP_ERROR(hipEventRecord(stop, NULL));
-      CHECK_HIP_ERROR(hipEventSynchronize(stop));
-      return std::vector{start, stop};
-    };
-
-    // Warmup
-    blas_benchmark::utils::warmup(blas_warmup);
-    CHECK_HIP_ERROR(hipStreamSynchronize(NULL));
+    alpha_rocm = alpha;
+    beta_rocm = beta;
+  }
 
-    blas_benchmark::utils::init_counters(state);
+#ifdef BLAS_VERIFY_BENCHMARK
+  // Reference gemm
+  std::vector<ref_scalar_t> c_ref(n * m, 0);
+  std::vector<scalar_t> c_temp(n * m, 0);
+
+  if constexpr (is_half) {
+    // Float-type variables for reference ops
+    ref_scalar_t alpha_f = alpha;
+    ref_scalar_t beta_f = beta;
+    std::vector<ref_scalar_t> a_f(m * k);
+    std::vector<ref_scalar_t> b_f(k * n);
+
+    // sycl::half to float reference type
```

Comment:
I think you forgot to remove the reference casting from the rocblas benchmarks.

Reply:
Yes will push these in a bit.